File: code/python/test/matrix/basic_matrix_multiplier_test.py
import unittest
from code.python.matrix.basic_matrix_multiplier import BasicMatrixMultiplier


class BasicMatrixMultiplierTest(unittest.TestCase):

    def setUp(self):
        self.A = [[1, 2], [3, 4]]
        self.B = [[5, 6], [7, 8]]
        self.expected = [[19, 22], [43, 50]]
        self.multiplier = BasicMatrixMultiplier()

    def test_multiply(self):
        result = self.multiplier.multiply(self.A, self.B)
        self.assert_matrix_almost_equal(self.expected, result, delta=1e-10)

    def test_generate_matrices_shape(self):
        n = 5
        A, B = self.multiplier.generate_matrices(n)
        self.assertEqual(len(A), n)
        self.assertEqual(len(A[0]), n)
        self.assertEqual(len(B), n)
        self.assertEqual(len(B[0]), n)

    def test_multiply_with_random_matrices(self):
        n = 3
        A, B = self.multiplier.generate_matrices(n)
        result = self.multiplier.multiply(A, B)
        self.assertEqual(len(result), n)
        self.assertEqual(len(result[0]), n)

    def assert_matrix_almost_equal(self, expected, actual, delta):
        self.assertEqual(len(expected), len(actual))
        for i in range(len(expected)):
            self.assertEqual(len(expected[i]), len(actual[i]))
            for j in range(len(expected[i])):
                self.assertAlmostEqual(expected[i][j], actual[i][j], delta=delta)

if __name__ == '__main__':
    unittest.main()

File: java/src/test/java/com/pabcablan/matrix/BasicMatrixMultiplierTest.java
package com.pabcablan.matrix;

import org.junit.Before;
import org.junit.Test;
import static org.junit.Assert.*;

public class BasicMatrixMultiplierTest {

    private double[][] A;
    private double[][] B;
    private double[][] expected;
    private BasicMatrixMultiplier multiplier;

    @Before
    public void setUp() {
        A = new double[][] {{1, 2}, {3, 4}};
        B = new double[][] {{5, 6}, {7, 8}};
        expected = new double[][] {{19, 22}, {43, 50}};
        multiplier = new BasicMatrixMultiplier();
    }

    @Test
    public void testMultiply() {
        double[][] result = multiplier.multiply(A, B);
        assertMatrixEquals(expected, result, 1e-10);
    }

    @Test
    public void testGenerateMatricesCreatesCorrectShape() {
        int n = 5;
        double[][][] matrices = multiplier.generateMatrices(n);
        assertEquals(n, matrices[0].length);
        assertEquals(n, matrices[0][0].length);
        assertEquals(n, matrices[1].length);
        assertEquals(n, matrices[1][0].length);
    }

    @Test
    public void testMultiplyWithRandomMatrices() {
        int n = 3;
        double[][][] matrices = multiplier.generateMatrices(n);
        double[][] result = multiplier.multiply(matrices[0], matrices[1]);
        assertEquals(n, result.length);
        assertEquals(n, result[0].length);
    }

    private void assertMatrixEquals(double[][] expected, double[][] actual, double delta) {
        assertEquals(expected.length, actual.length);
        for (int i = 0; i < expected.length; i++) {
            assertEquals(expected[i].length, actual[i].length);
            for (int j = 0; j < expected[i].length; j++) {
                assertEquals(expected[i][j], actual[i][j], delta);
            }
        }
    }
}